File: server/machines.js
import { GOODS, MILL_REQUIREMENT, SPRINKLER } from '../shared/materials.js';

const MATERIAL_IDS = Object.keys(MILL_REQUIREMENT);
const QUALITY_ORDER = ['C', 'B', 'A', 'A+']; // Milling consumes the lowest grade first

/**
 * Reads the persisted mill status straight from storage without requiring a
 * MachinesManager instance (used by contract generation, which may run before
 * machines state has been normalized).
 */
export function isMillRestoredInState(state) {
  return state?.machines?.mill?.status === 'restored';
}

/**
 * The Market Court machine shop. Machines have restore/consume lifecycles
 * rather than prices, so this lives beside gardens.js instead of economy.js.
 * All economically meaningful mutations are validated here on the server and
 * persisted to storage.state.machines = { mill: {...} }.
 */
export class MachinesManager {
  constructor(storage) {
    this.storage = storage;
    const saved = this.storage.state.machines;
    this.mill = saved && saved.mill && typeof saved.mill === 'object'
      ? this.normalizeMill(saved.mill)
      : this.createDefaultMill();
    this.persist();
  }

  createDefaultMill() {
    return {
      status: 'broken',
      required: { ...MILL_REQUIREMENT },
      contributed: { copper: 0, timber: 0, glass: 0 },
      restoredAt: null,
    };
  }

  normalizeMill(mill) {
    const base = this.createDefaultMill();
    mill.status = mill.status === 'restored' ? 'restored' : 'broken';
    if (!mill.required || typeof mill.required !== 'object') mill.required = { ...base.required };
    if (!mill.contributed || typeof mill.contributed !== 'object') mill.contributed = { ...base.contributed };
    for (const id of MATERIAL_IDS) {
      const required = Number.isFinite(Number(mill.required[id])) && Number(mill.required[id]) > 0
        ? Math.floor(Number(mill.required[id]))
        : base.required[id];
      const contributed = Number(mill.contributed[id]);
      mill.required[id] = required;
      mill.contributed[id] = Number.isFinite(contributed) && contributed > 0 ? Math.floor(contributed) : 0;
    }
    if (mill.restoredAt !== null && !Number.isFinite(Number(mill.restoredAt))) mill.restoredAt = null;
    return mill;
  }

  persist() {
    this.storage.state.machines = { mill: this.mill };
    this.storage.save();
  }

  getStatus() {
    return {
      mill: {
        status: this.mill.status,
        required: { ...this.mill.required },
        contributed: { ...this.mill.contributed },
        restoredAt: this.mill.restoredAt,
      },
    };
  }

  isMillRestored() {
    return this.mill.status === 'restored';
  }

  remainingNeed(material) {
    const required = this.mill.required[material];
    if (!Number.isFinite(required)) return 0;
    return Math.max(0, required - (this.mill.contributed[material] || 0));
  }

  /**
   * Applies a material contribution atomically: the request is clamped to
   * both the player's stock and the mill's remaining need, materials leave
   * the inventory only when at least one unit is accepted, and restoration
   * completes in the same step that receives the final unit. Runs
   * synchronously, so concurrent messages are naturally serialized.
   */
  contribute(player, material, quantity, now = Date.now()) {
    if (this.mill.status !== 'broken') return { success: false, reason: 'mill_already_restored' };
    if (!(material in this.mill.required)) return { success: false, reason: 'material_not_needed' };

    const requested = Math.floor(Number(quantity));
    if (!Number.isFinite(requested) || requested <= 0) {
      return { success: false, reason: 'invalid_quantity' };
    }

    if (!player.materials || typeof player.materials !== 'object') player.materials = {};
    const held = Math.floor(Number(player.materials[material]) || 0);
    const remaining = this.remainingNeed(material);
    if (remaining <= 0) return { success: false, reason: 'material_fulfilled' };
    if (held <= 0) return { success: false, reason: 'insufficient_materials' };

    const applied = Math.min(requested, held, remaining);
    player.materials[material] = held - applied;
    if (player.materials[material] <= 0) delete player.materials[material];
    this.mill.contributed[material] += applied;

    let restored = false;
    if (MATERIAL_IDS.every(id => this.remainingNeed(id) <= 0)) {
      this.mill.status = 'restored';
      this.mill.restoredAt = now;
      restored = true;
    }
    this.persist();

    return { success: true, material, applied, restored, machine: this.getStatus() };
  }

  /**
   * Milled goods: a restored mill converts wheat into flour 1:1 (no quality
   * grades in this slice), consuming the player's lowest-grade wheat first.
   */
  millWheat(player, quantity = 1) {
    if (!this.isMillRestored()) return { success: false, reason: 'mill_broken' };

    const requested = Math.floor(Number(quantity));
    if (!Number.isFinite(requested) || requested <= 0) {
      return { success: false, reason: 'invalid_quantity' };
    }
    if (!player.inventory || typeof player.inventory.produce !== 'object') {
      return { success: false, reason: 'no_wheat' };
    }

    let available = 0;
    for (const quality of QUALITY_ORDER) {
      available += player.inventory.produce[`wheat_${quality}`] || 0;
    }
    if (available <= 0) return { success: false, reason: 'no_wheat' };

    const milled = Math.min(requested, available);
    let leftToTake = milled;
    for (const quality of QUALITY_ORDER) {
      if (leftToTake <= 0) break;
      const key = `wheat_${quality}`;
      const take = Math.min(player.inventory.produce[key] || 0, leftToTake);
      if (take <= 0) continue;
      player.inventory.produce[key] -= take;
      if (player.inventory.produce[key] <= 0) delete player.inventory.produce[key];
      leftToTake -= take;
    }

    player.inventory.produce[`${GOODS.flour.id}_B`] = (player.inventory.produce[`${GOODS.flour.id}_B`] || 0) + milled;
    return { success: true, milled, good: GOODS.flour };
  }

  /**
   * Crafts a garden fixture from gathered materials. The sprinkler kit is
   * added to the player's inventory; placement onto a bed tile is a separate
   * validated garden action.
   */
  craft(player, fixture) {
    if (fixture !== SPRINKLER.id) return { success: false, reason: 'unknown_fixture' };
    if (!player.materials || typeof player.materials !== 'object') player.materials = {};

    for (const [material, cost] of Object.entries(SPRINKLER.cost)) {
      const held = Math.floor(Number(player.materials[material]) || 0);
      if (held < cost) return { success: false, reason: 'insufficient_materials', material };
    }
    for (const [material, cost] of Object.entries(SPRINKLER.cost)) {
      player.materials[material] -= cost;
      if (player.materials[material] <= 0) delete player.materials[material];
    }

    if (!player.inventory || typeof player.inventory !== 'object') player.inventory = {};
    player.inventory.sprinklers = (Math.floor(Number(player.inventory.sprinklers)) || 0) + 1;

    return { success: true, fixture: SPRINKLER.id, name: SPRINKLER.name };
  }
}

File: server_elixir/lib/afterlight/restoration/machines.ex
defmodule Afterlight.Restoration.Machines do
  @moduledoc """
  Mill restoration, milling, and sprinkler crafting from `server/machines.js`.
  """

  alias Afterlight.Economy.Catalog
  alias Afterlight.Parity.Numeric
  alias Afterlight.Restoration.Materials

  @quality_order ["C", "B", "A", "A+"]
  @flour Catalog.get("flour")

  def fresh_state do
    mill = default_mill()
    %{"storage" => %{"state" => %{"machines" => %{"mill" => mill}}}, "mill" => mill}
  end

  def mill_restored_in_state?(state) do
    get_in(state || %{}, ["machines", "mill", "status"]) == "restored"
  end

  def default_mill do
    %{
      "status" => "broken",
      "required" => Materials.mill_requirement(),
      "contributed" => %{"copper" => 0, "timber" => 0, "glass" => 0},
      "restoredAt" => nil
    }
  end

  def snapshot(mill) do
    %{
      "status" => mill["status"],
      "required" => mill["required"],
      "contributed" => mill["contributed"],
      "restoredAt" => mill["restoredAt"]
    }
  end

  def contribute(state, player, material, quantity, now) do
    mill = state["mill"]

    cond do
      mill["status"] != "broken" ->
        {state, %{"success" => false, "reason" => "mill_already_restored"}}

      not Map.has_key?(mill["required"], material) ->
        {state, %{"success" => false, "reason" => "material_not_needed"}}

      true ->
        case Numeric.js_to_number(quantity) do
          :nan ->
            {state, %{"success" => false, "reason" => "invalid_quantity"}}

          n ->
            requested = js_floor(n)

            if requested <= 0 do
              {state, %{"success" => false, "reason" => "invalid_quantity"}}
            else
              held = js_held(player, material)
              remaining = remaining_need(mill, material)

              cond do
                remaining <= 0 ->
                  {state, %{"success" => false, "reason" => "material_fulfilled"}}

                held <= 0 ->
                  {state, %{"success" => false, "reason" => "insufficient_materials"}}

                true ->
                  applied = Enum.min([requested, held, remaining])
                  mill = put_in(mill, ["contributed", material], mill["contributed"][material] + applied)

                  {restored, mill} =
                    if Enum.all?(Materials.material_ids(), &(remaining_need(mill, &1) <= 0)) do
                      {true, mill |> Map.put("status", "restored") |> Map.put("restoredAt", now)}
                    else
                      {false, mill}
                    end

                  state = store_mill(state, mill)

                  {state,
                   %{
                     "success" => true,
                     "material" => material,
                     "applied" => applied,
                     "restored" => restored,
                     "machine" => %{"mill" => snapshot(mill)}
                   }}
              end
            end
        end
    end
  end

  def mill_wheat(state, player, quantity) do
    mill = state["mill"]

    if mill["status"] != "restored" do
      {state, %{"success" => false, "reason" => "mill_broken"}}
    else
      case Numeric.js_to_number(quantity) do
        :nan ->
          {state, %{"success" => false, "reason" => "invalid_quantity"}}

        n ->
          requested = js_floor(n)

          if requested <= 0 do
            {state, %{"success" => false, "reason" => "invalid_quantity"}}
          else
            produce =
              case get_in(player || %{}, ["inventory", "produce"]) do
                %{} = produce -> produce
                _ -> nil
              end

            if produce == nil do
              {state, %{"success" => false, "reason" => "no_wheat"}}
            else
              available =
                Enum.sum(Enum.map(@quality_order, &Map.get(produce, "wheat_#{&1}", 0)))

              if available <= 0 do
                {state, %{"success" => false, "reason" => "no_wheat"}}
              else
                milled = min(requested, available)

                produce =
                  Enum.reduce(@quality_order, {produce, milled}, fn quality, {produce, left} ->
                    if left <= 0 do
                      {produce, 0}
                    else
                      key = "wheat_#{quality}"
                      have = Map.get(produce, key, 0)
                      take = min(have, left)

                      if take <= 0 do
                        {produce, left}
                      else
                        left_after = have - take
                        produce = if left_after <= 0, do: Map.delete(produce, key), else: Map.put(produce, key, left_after)
                        {produce, left - take}
                      end
                    end
                  end)
                  |> elem(0)

                produce = Map.put(produce, "flour_B", Map.get(produce, "flour_B", 0) + milled)
                state = store_inventory_produce(state, player, produce)

                {state, %{"success" => true, "milled" => milled, "good" => @flour}}
              end
            end
          end
      end
    end
  end

  def craft(state, player, fixture) do
    if fixture != "sprinkler" do
      {state, %{"success" => false, "reason" => "unknown_fixture"}}
    else
      materials = Map.get(player || %{}, "materials") || %{}

      failure =
        Enum.find(Materials.sprinkler_cost(), fn {material, cost} ->
          js_held(%{"materials" => materials}, material) < cost
        end)

      case failure do
        {material, _cost} ->
          {state, %{"success" => false, "reason" => "insufficient_materials", "material" => material}}

        nil ->
          {state, %{"success" => true, "fixture" => "sprinkler", "name" => "Garden Sprinkler"}}
      end
    end
  end

  defp remaining_need(mill, material) do
    required = mill["required"][material]

    if is_number(required) do
      max(0, required - (mill["contributed"][material] || 0))
    else
      0
    end
  end

  defp store_mill(state, mill) do
    state
    |> put_in(["storage", "state", "machines", "mill"], mill)
    |> Map.put("mill", mill)
  end

  defp store_inventory_produce(state, _player, _produce), do: state

  defp js_floor(n) when is_integer(n), do: n
  defp js_floor(n) when is_float(n), do: Float.floor(n) |> trunc()

  defp js_held(player, material) do
    materials = Map.get(player || %{}, "materials") || %{}

    case Numeric.js_to_number(Map.get(materials, material, :absent)) do
      :nan -> 0
      n -> js_floor(n)
    end
  end
end
